### PR TITLE
In Which We Add Support for Sample Weights to MLPs

### PR DESCRIPTION
**Pull request by waltaskew**
- support sample weights in MLPClassifier
- support sample weights in MLPRegressor
- add tests ensuring sample weights are handled reasonably

@mheilman

### CODE REVIEW

**Comment by beckermr** [comment]
A quick comment here. We should normalize the losses by the sum of the sample weights in order to keep them on a decent numerical scale.

File: muffnn/mlp/mlp_classifier.py [hunk 1]
```diff
@@ -129,19 +129,23 @@ def _init_model_objective_fn(self, t):
         if self.multilabel_:
             cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(
                 logits=t, labels=tf.cast(self.input_targets_, np.float32))
+            cross_entropy = tf.multiply(cross_entropy, self._sample_weight)
```
Comment (beckermr) [inline muffnn/mlp/mlp_classifier.py]: Does this broadcast correctly?

For multilabel, I think the shape of of cross_entropy is [n_examples, n_classes] and the weights are [n_examples]. You might need to add a trailing dimensions to get this to work?

Comment (waltaskew) [reply muffnn/mlp/mlp_classifier.py]: It does not broadcast correctly. The fact that the tests passed makes me worry how well we're testing the multi-label case.

File: muffnn/mlp/tests/util.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    sample_weight = np.ones(X_train.shape[0])
+    clf1 = make_classifier_func().fit(X_train, y_train)
+    clf2 = make_classifier_func().fit(
+        X_train, y_train,
+        sample_weight=sample_weight,
+    )
+    assert_clfs_almost_equal(clf1, clf2)
+
+    # Fitting twice with half sample-weights should result
+    # in same classifier as fitting once with full weights.
```
Comment (mheilman) [inline muffnn/mlp/tests/util.py]: hmm... i don't think running for twice as many epochs with sample weights scaled by 0.5 will lead to the same model for non-vanilla SGD.  Also, shouldn't the models be the same after a single fit of clf2 because of the batch sample weight normalization?

Does this fail if partial_fit is only called once below for clf2?

File: muffnn/mlp/tests/util.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    assert_clfs_almost_equal(clf1, clf2)
+
+    # Duplicated inputs should have the same result as increasing
+    # the sample weights to the same degree as the duplication.
+    ind = rng.randint(0, X_train.shape[0], X_train.shape[0])
+    sample_weight = np.bincount(ind, minlength=X_train.shape[0])
+
+    clf1 = make_classifier_func().fit(X_train[ind], y_train[ind])
+    clf1 = make_classifier_func().fit(
+        X_train, y_train, sample_weight=sample_weight
```
Comment (mheilman) [inline muffnn/mlp/tests/util.py]: does this fail if you don't include the sample weights here (after fixing the copy-paste error)?

File: muffnn/mlp/tests/util.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    )
+    assert_clfs_almost_equal(clf1, clf2)
+
+    # Duplicated inputs should have the same result as increasing
+    # the sample weights to the same degree as the duplication.
+    ind = rng.randint(0, X_train.shape[0], X_train.shape[0])
+    sample_weight = np.bincount(ind, minlength=X_train.shape[0])
+
+    clf1 = make_classifier_func().fit(X_train[ind], y_train[ind])
+    clf1 = make_classifier_func().fit(
```
Comment (mheilman) [inline muffnn/mlp/tests/util.py]: copy-paste error?  i think this should be clf2.

why didn't that cause a test failure?

File: muffnn/mlp/mlp_classifier.py (hunk 1 above)
Comment (beckermr) [reply muffnn/mlp/mlp_classifier.py]: Agreed. Sam and I worked on those tests a lot but I am not going to claim they are bullet proof.

File: muffnn/mlp/tests/util.py (hunk 4 above)
Comment (waltaskew) [reply muffnn/mlp/tests/util.py]: It doesn't cause a failure because all the clf1 and clf2s are learning the same model. It's comparing against the wrong clf2, but the wrong clf2 is still trained on the same data.

The window for determining if two classifiers make the same prediction may be too large, though.

File: muffnn/mlp/tests/util.py (hunk 2 above)
Comment (waltaskew) [reply muffnn/mlp/tests/util.py]: Yeah, I don't think this test makes sense because we're normalizing by the sum of the weights.
My original thinking was to shrink the loss with the weights to force the training to take smaller steps, but normalization means this isn't happening -- I'll remove this test.

**Comment by waltaskew** [comment]
> My main comment here is that the tests seem a bit excessive. It seems sufficient to test that the weights are handled properly in a simple technical sense, as opposed to testing our assumptions about how the weights should work. To this end, only the last test in the assert function in the test suite seems relevant.

I'd like to keep the test asserting that not passing sample weights give the same result as passing sample weights of all 1s, because that seems like a reasonable invariant that could catch errors in handling absent sample weights.

I'm removing the two partial fits test, which leaves the test asserting that duplicated inputs are equivalent to up-weighted inputs. I don't have particular allegiances to that test.

**Comment by waltaskew** [comment]
I improved the docs and comments and removed the brittle tests. Anything else?

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/)
 and this project adheres to [Semantic Versioning](http://semver.org/).
 
+## Unreleased
+
+- Added support for the `sample_weight` keyword argument to the `fit`
```
Comment (mheilman) [inline CHANGELOG.md]: Could you put this under an " ### Added" section please?

File: muffnn/mlp/base.py [hunk 6]
```diff
@@ -66,6 +66,15 @@ def fit(self, X, y, monitor=None):
             The monitor can be used for various things such as computing
             held-out estimates, early stopping, model introspection,
             and snapshoting.
+        sample_weight : numpy array of shape [n_samples,]
+            Per-sample weights. Re-scale the loss per sample.
+            Higher weights force the estimator to put more emphasis
+            on these samples.
+            Sample weights are normalized per-batch. As the batch size
```
Comment (mheilman) [inline muffnn/mlp/base.py]: I think just "Sample weights are normalized per-batch." is fine here.

File: muffnn/mlp/mlp_classifier.py [hunk 7]
```diff
@@ -163,6 +183,15 @@ def partial_fit(self, X, y, monitor=None, classes=None):
             The monitor can be used for various things such as computing
             held-out estimates, early stopping, model introspection,
             and snapshoting.
+        sample_weight : numpy array of shape (n_samples,)
+            Per-sample weights. Re-scale the loss per sample.
+            Higher weights force the estimator to put more emphasis
+            on these samples.
+            Sample weights are normalized per-batch. As the batch size
```
Comment (mheilman) [inline muffnn/mlp/mlp_classifier.py]: see comment above

**Comment by beckermr** [comment]
LGTM from me modulo mike’s comments. Thanks a bunch for doing this Walt! (Hammer) (huge fan)

**Comment by waltaskew** [comment]
CHANGELOG changed and docstrings shortened, anything else?
